053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+  the Elasticsearch cluster. For a `Postgres` implementation, `Setup()` will
+  check if the `build_events` table exists, and if not, create it (note: this
+  was previously done by DB migrations - more on this in [Migrations](#migrations))
+
+* `Initialize(build db.Build)` will be triggered when a build is first created, and
+  before any build events are `Put` into the store. For the `Postgres`
+  implementation, this will create a sequence like `build_event_id_seq_x` for the
+  current build. It will also create either the table `pipeline_build_events_x`
+  or `team_build_events_x`, depending on whether it's a one-off build or not.
+  This represents a behaviour change from what we have now, where this table is
+  created when the team/pipeline is first created, but it shouldn't really have
+  much impact.
+
+* `Finalize(build db.Build)` will be triggered when a build is `Finish`ed. No
+  build events may be `Put` into the store after this is called. For
+  `Postgres`, this involves dropping the `build_event_id_seq_x` sequence.
+
+* `Put(build db.Build, events []atc.Event)` is called whenever
+  `build.SaveEvent(event)` is called in order to write build events to the
+  external store. The reason for accepting a list of `events` (despite
+  `db.Build.SaveEvent` only ever passing in a single event) is because if we
+  want to migrate build events to another `EventStore`, it would be super slow
+  making make a huge number of small `Put` calls when we can do one batch `Put`
+  per Build.
+
+* `Get(build db.Build, requested int, cursor *Key)` fetches events from the
+  `EventStore`, starting from an initial `Key` (which is excluded from the
+  result set - e.g.  if `cursor` points to the 10th `Key`, the 11th event will be
+  the next one fetched). `Key` can be any type, and is implementation-specific.
+  For `Postgres`, this will likely be something like `uint`, but other stores
+  (e.g. DynamoDB) that don't provide easy support for auto-incrementing IDs may
+  choose a different type of `Key`. Regardless of the type of `Key`, if the goal
+  is to fetch all events, `cursor` should initially point to `nil` (but the
+  pointer itself must not be `nil`). For instance, you can iterate over all
+  events for a completed build like so:
+
+  ```go
+  var cursor Key
+  batchSize := 1000
+  for {
+      events, err := eventStore.Get(build, batchSize, &cursor)
+      if err != nil {
+          panic(err)
+      }
+      // process events here
+      if len(events) < batchSize {
+          // there are no more events (see note below about `requested`)
+          break
+      }
+  }
+  ```
+
+  However, when possible, consumers of these build events should prefer to use
+  `db.Build.Events()`, which queries the `EventStore` and subscribes to a
+  notification bus to allow tracking in-progress builds (more on this in
+  [Notifications](#notifications)).
+
+  `Get` also takes in `requested` number of events to fetch. This is a soft
```

Comment:
pls bear with me, but I got a bit confused with the wording here :sweat_smile:  so, is `requested` meant to be the maximum number of events that would come back, regardless of how many events are there?

e.g., a [`read(2)`](http://man7.org/linux/man-pages/man2/read.2.html) with a `count = 1024` means that it'll never try to fill my buffer with more than `1024` bytes - is this what you're trying to achieve?

Reply:
Yeah, I was kinda struggling with how to word this.

The intention was actually the opposite of `read(2)`. The use case I was trying to account for that sometimes it's difficult to fetch exactly `requested` events from the store. Suppose  we have a Cold Storage implementation of `EventStore` (all events for a build are stored in a single file). If `requested = 1000` and there are 1500 events for the build, it doesn't really make sense to make two `Gets` and have to download the file from Cold Storage twice. Even though the client only `requested` 1000 events, since we've already fetched all 1500, we might as well give them everything.

Granted, another way to approach this would be for the Cold Storage implementation to only return 1000 events, but keep track of the remaining 500 events for the next time `Get` is called so we don't have to download the file again. Not sure what's approach is better.

I agree that with `read(2)`, knowing that you'll never try to fill a buffer with more than `count` bytes is a good guarantee, but given that the client isn't providing a buffer to fill, my thinking was exceeding `requested` isn't dangerous (let me know if you disagree!)

Reply:
So, to summarize, `requested` isn't an upper limit on the number of events to return.

If there are at least `requested` events in the event store for a particular build, then the implementation can decide whether it makes sense to return exactly `requested`, or more.

e.g. in the example above, the Cold storage implementation would return all 1500 events, despite `requested == false`. It would be equally valid to first return 1000 and then return 500, or even to return 1200 and then 300.